### DISCUSSION TITLE
Use more actual zookeeper version

### DISCUSSION THREAD

**Discussion by AydinChavez** [strimzi/strimzi-kafka-operator#7528] [Q&A]
Hi Jakub,

is it somehow possible to manually link/inject a newer zookeeper image (e.g. 3.8.0) into the kafka-operator?

I am asking because the older ones use log4j and due to security concerns (some CVEs attached to 3.6.*) we would like to incorporate a newer version.

If it is possible, what would be the easiest approach from deployment perspective?

**Comment by scholzj** [comment]
Strimzi uses the ZooKeeper which is part of Apache Kafka and its binaries. One reason is that it is easier. Other one is that it is well tested and proven since it is part of Kafka itself. TBH, I'm not sure if Kafka has some special reason why it still uses ZooKeeper 3.6. Maybe nobody wants to update it anymore with the ZooKeeper removal slowly progressing towards the finale.

**Reply by AydinChavez** [reply]
Okay thank you. Maybe I should consider using kraft mode than, since I think it is now production ready.

**Reply by scholzj** [reply]
Well, not in Strimzi. Even in Kafka I think it is quite questionable what _production ready_ means.

**Reply by AydinChavez** [reply]
You mean Kraft mode is not supported in strimzi?

**Reply by scholzj** [reply]
There is an experimental feature gate for it which you can use for example for development or testing. But there is not yet any production ready support. And I expect we will support it when it is actually production ready - feature complete, with upgrade and migration from ZooKeeper based Kafka etc. Right now, we do not have a capacity to support different paths for KRaft and ZooKeeper based Kafka and it would be also quite hard to guarantee that you can properly upgrade in the future etc.

**Reply by AydinChavez** [reply]
Understood, thank you. 

Coming back to my first question - I would like to build and incorporate an own kafka image for the kafka-operator with zookeeper 3.8.0 which is backwards compatible to clients using zk >= 3.6.0

I guess https://github.com/strimzi/strimzi-kafka-operator/blob/main/helm-charts/helm3/strimzi-kafka-operator/values.yaml is the place where I would set it in case I deploy via helm? 

Do you have any hint how to build the strimzi operator with a custom kafka image locally? Is it tough or a kind of a two-liner? :) (I mean building strimzi operator in particular, not building docker images).

**Reply by scholzj** [reply]
It depends on what exactly you have and want to customize. The images which will be used can be configured in the Cluster Operator deployment. So that is not an problem.

As for building your own image, I guess you have two options:
* You can modify the [kafka-versions.yaml](https://github.com/strimzi/strimzi-kafka-operator/blob/main/kafka-versions.yaml) file which defines from where should the Kafka binary be pulled.
* Or you can write your own Dockerfile which basically takes the existing image and replaces the JARs with your JARs and then build and use this image.

To be clear, I'm not aware of anyone doing that -> so these are the general ideas. You would need to try them.

I have also no idea what impact the new ZooKeeper version might have on upgrades etc. In the past, ZooKeeper upgrades between minor versions were not always smooth.
